### ISSUE TITLE
Akka Streams (Websockets) appears to be leaking memory

### ISSUE THREAD

**Issue by strelec** [playframework/playframework#7723]
In my web app, I observe about +50MB of usage per day. There appears to be a huge number of Concat objects in my state dump.

Instance Counts for All Classes (excluding platform)

353102 instances of class akka.stream.impl.Concat 
352356 instances of class scala.collection.immutable.$colon$colon 
205317 instances of class akka.stream.impl.fusing.GraphInterpreter$Connection 
198116 instances of class [Lakka.stream.impl.fusing.GraphInterpreter$Connection; 
162219 instances of class akka.stream.Attributes 
100913 instances of class akka.stream.Outlet 
93703 instances of class akka.stream.Inlet 
82209 instances of class scala.collection.immutable.Vector 
79268 instances of class akka.stream.impl.fusing.GraphStageModule 
75665 instances of class akka.stream.impl.AtomicTraversalBuilder 
68461 instances of class akka.stream.impl.MaterializeAtomic 
57653 instances of class akka.stream.impl.PushAttributes 
50445 instances of class akka.stream.FlowShape 
45325 instances of class akka.http.impl.engine.parsing.HttpHeaderParser$ValueBranch 
28816 instances of class akka.stream.scaladsl.Broadcast$$anon$4$$anon$15 
25214 instances of class akka.stream.impl.fusing.Collect$$anon$2 
25214 instances of class akka.stream.impl.fusing.Collect$$anon$2$$Lambda$783 
23577 instances of class akka.stream.stage.GraphStageLogic$$anon$1 
21612 instances of class scala.collection.LinearSeqLike$$anon$1 
18961 instances of class akka.http.scaladsl.model.headers.ProductVersion 
18036 instances of class akka.actor.ChildActorPath 
18029 instances of class akka.actor.ActorCell 
18012 instances of class akka.stream.impl.fusing.Collect 
18011 instances of class akka.stream.impl.fusing.Map$$anon$7 
18010 instances of class akka.stream.impl.fusing.GraphStages$Identity$$anon$1

**Comment by strelec** [comment]
This is Play 2.6.1.

**Comment by richdougherty** [comment]
Thanks for the report.

- Are you using an Akka Streams APIs directly in your project or is this just from using regular Play APIs? If your app is just using regular Play APIs then there shouldn't be a leak unless we have a bug in Play's usage of Akka Streams.
- I guess these are numbers from your live production system. Is there any chance you could test it out on a simple project so we can run locally to reproduce it? You could use _wrk_ or _ab_ to create load.
- If you do get the chance it would be good to check on Play 2.6.2 also, although I don't believe we've changed much about how we use Akka Streams since Play 2.6.1.

**Comment by strelec** [comment]
1. I am not using any Akka Streams APIs directly.
2. I do not have a scenario or resources to reproduce it locally. I get a lot of concurrent websocket connections in production. I do have a heap dump that is 300 megs in size.
3. I will try and report back.

**Comment by richdougherty** [comment]
Thanks for the extra information @strelec. How many concurrent connections would you say you had active when you got those object counts?

The reason I ask for the number of concurrent connections is that it allows you to work out how many object we're seeing per connection. E.g. if there are about 18,000 connections active, then we can see about 20 Concat objects per connection. Then we can ask ourselves if that number seems reasonable or if it looks too high.

**Comment by richdougherty** [comment]
Ccing @ktoso @jrudolph from the akka-http team. Hi guys!

**Comment by strelec** [comment]
The heap dump was taken at night, when connections drop to 0. So this really is stale data.

**Comment by ktoso** [comment]
Would you be able to prepare some minimized reproducer?
The above just says that there's plenty streams - it could be from any place.
Could you pinpoint where the references are being held if you have the full heapdump at hand? (profilers can help with locating that)

**Comment by jrudolph** [comment]
> The heap dump was taken at night, when connections drop to 0.

Even if you cannot minimize it, could you share a heap dump from when the connections drop to 0?

**Comment by richdougherty** [comment]
@jrudolph, I think @strelec's heap dump is already from when the connections drop to 0.

**Comment by jrudolph** [comment]
Could we get access to it?

**Comment by richdougherty** [comment]
@strelec, if you want to keep the heap dump private you could email me a link to download it. My email address is rich.dougherty@lightbend.com. We can keep it confidential.

**Comment by jrudolph** [comment]
Would be also nice to know the exact dependencies. If @strelec uses vanilla Play 2.6.1, what akka-http/akka-stream version would that translate to?

**Comment by richdougherty** [comment]
Here you go!

https://github.com/playframework/playframework/blob/2.6.1/framework/project/Dependencies.scala#L11-L12

**Comment by jrudolph** [comment]
Thanks, @richdougherty. So, having the memory dump or a reproducer would help a lot here.

**Comment by strelec** [comment]
https://drive.google.com/file/d/0B4UFCNSTmGhGa1hnSXNmOVE5a3c/view?usp=sharing

This is the dump.

**Comment by richdougherty** [comment]
Thanks. I see about 3,602 WebSocketHandlers, so I think that's how many connections Play is keeping open at the time. That could be correct or it could be a leak. Can I just check how you confirmed there were no active connections. Did you run _netstat_ or something like that?

**Comment by jrudolph** [comment]
Thanks a lot @strelec for sharing. I don't see as many sockets so the leak could be real.

**Comment by strelec** [comment]
There were none

I also suspected those are the sockets that have not been properly closed. This is definitely something new with play 2.6.* versions, I have not had any such problems before.

**Comment by jrudolph** [comment]
I already found https://github.com/akka/akka/issues/23535 which isn't the root cause but may contribute to it.

**Comment by jrudolph** [comment]
The part of the websocket stream that stays open has this form:

```
(akka-http) FrameEventParser -> (play) WebsocketHandler1 -> (play) WebsocketHandler2 -> Broadcast 1
Broadcast1 -> Collect -> (play) WebsocketFlowHandler
Broadcast1 => Broadcast2 => IgnoreSink

-> connection that waits for demand, i.e. that waits for the downstream to pull
=> connection that waits for data, i.e. that was pulled and waits for a push
```

The problematic stages are thus:
 * akka-http `FrameEventParser`: does not report completion if not pulled. Hard to say if you should call this a bug, it's more a detail of the current implementation (https://github.com/akka/akka/issues/23536).
 * play `WebsocketFlowHandler`: keeps waiting for completion on the frame input port even if all other ports have been closed. It might be that the bug is in [this line](https://github.com/playframework/playframework/blob/master/framework/src/play-server/src/main/scala/play/core/server/common/WebSocketFlowHandler.scala#L188). The comment above says "complete both remote out and remote in" but then does `cancel(appIn)` instead of `cancel(remoteIn)`. Probably the stage could just call `completeStage` after sending out the corresponding close frame.

This "deadlock" can be solved on both ends.

**Comment by schmitch** [comment]
that part did I wrote, cause with completeStage() some netty websockets might fail:  https://github.com/playframework/playframework/commit/dd8b84fa1125c90c909dd3cb4355c8310c134bc4#diff-745ef9331ba1901efedd1607f6675eea

**Comment by jrudolph** [comment]
Another preliminary fix for play could be to add a simple `buffer(1)` stage after the akka-http `FrameEventParser`. That should make sure that completion is reported eagerly as well.

**Comment by schmitch** [comment]
@strelec is this still the case with newer akka versions?

**Comment by dkarivalis** [comment]
i believe this was fixed in https://github.com/playframework/playframework/pull/8292

no longer seeing a memory leak when using WebSockets after switching to Play's nightly build

**Comment by schmitch** [comment]
well this is also in 2.6.13 https://github.com/playframework/playframework/commit/437c23544b4b584344372e892bcbbaca9857f0d1

**Comment by dkarivalis** [comment]
> well this is also in 2.6.13 437c235

ah i wasn't aware it made it to a patch version. thanks @schmitch !

**Comment by ktoso** [comment]
Excellent, thanks for checking and reporting back!

**Comment by marcospereira** [comment]
This was fixed by #8292 as confirmed by @dkarivalis. :-)
